pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/impl/LeastResourceUsageWithWeight.java

```diff
@@ -0,0 +1,167 @@
+        Double historyUsage = brokerAvgResourceUsageWithWeight.get(broker);
+        double resourceUsage = brokerData.getLocalData().getMaxResourceUsageWithWeight(
+                conf.getLoadBalancerCPUResourceWeight(),
+                conf.getLoadBalancerMemoryResourceWeight(),
+                conf.getLoadBalancerDirectMemoryResourceWeight(),
+                conf.getLoadBalancerBandwithInResourceWeight(),
+                conf.getLoadBalancerBandwithOutResourceWeight());
+        historyUsage = historyUsage == null
+                ? resourceUsage : historyUsage * historyPercentage + (1 - historyPercentage) * resourceUsage;
+        if (log.isDebugEnabled()) {
+            log.debug(
+                    "Broker {} get max resource usage with weight: {}, history resource percentage: {}%, CPU weight: "
+                            + "{}, MEMORY weight: {}, DIRECT MEMORY weight: {}, BANDWIDTH IN weight: {}, BANDWIDTH "
+                            + "OUT weight: {} ",
+                    broker, historyUsage, historyPercentage, conf.getLoadBalancerCPUResourceWeight(),
+                    conf.getLoadBalancerMemoryResourceWeight(), conf.getLoadBalancerDirectMemoryResourceWeight(),
+                    conf.getLoadBalancerBandwithInResourceWeight(),
+                    conf.getLoadBalancerBandwithOutResourceWeight());
+        }
+        brokerAvgResourceUsageWithWeight.put(broker, historyUsage);
+        return historyUsage;
+    }
+
+    /**
+     * Find a suitable broker to assign the given bundle to.
+     * This method is not thread safety.
+     *
+     * @param candidates     The candidates for which the bundle may be assigned.
+     * @param bundleToAssign The data for the bundle to assign.
+     * @param loadData       The load data from the leader broker.
+     * @param conf           The service configuration.
+     * @return The name of the selected broker as it appears on ZooKeeper.
+     */
+    @Override
+    public Optional<String> selectBroker(Set<String> candidates, BundleData bundleToAssign, LoadData loadData,
+                                         ServiceConfiguration conf) {
+        bestBrokers.clear();
+        // Maintain of list of all the best scoring brokers and then randomly
+        // select one of them at the end.
+        double totalUsage = 0.0d;
+        for (String broker : candidates) {
+            BrokerData brokerData = loadData.getBrokerData().get(broker);
+            double usageWithWeight = getMaxResourceUsageWithWeight(broker, brokerData, conf);
+            totalUsage += usageWithWeight;
+        }
+
+        final double avgUsage = totalUsage / candidates.size();
+        final double diffThreshold =
+                conf.getLoadBalancerAverageResourceUsageDifferenceThresholdShedderPercentage() / 100.0;
+        brokerAvgResourceUsageWithWeight.forEach((broker, avgResUsage) -> {
+            if (avgResUsage + diffThreshold <= avgUsage) {
```

Comment:
I think this algo is better than the `minScore`, but we still have the `small randomization pool` issue here.
example) b1:4, b2:20, b3:20

Here, still, b1 will be repeatedly selected until b1's load is updated to the leader, though all of these brokers should be candidates as their loads are low. (As we discussed, when many bundles are unloading in a short period, this might lead to b1 ending up overloaded, like b1:90%, b2:20, b3:20) 

Can we do a high_load_threshold_filter and random selection?

I think one could implement a usage-and-selection-weighted random selection to weigh more on the underloaded brokers.

Reply:
You are right. The original intention of this algorithm is to select a larger `randomization pool`. It doesn't work well for small clusters. But there is a compromise improvement point, that is, adjusting the `diffThreshold` to a negative number, which can expand the `randomization pool`.
 (eg: b1:4, b2:20, b3:20, set `diffThreshold`=-15, all of brokers can be candidates.) 
This `diffThreshold` parameter takes effect dynamically, so it can be adjusted dynamically according to the status of cluster. Do you have a better suggestion? :)

Reply:
@heesung-sn @dragonls PTAL

Reply:
Please explain the negative value in the doc, In the config definition.

Another edge case is that with the negative threshold, it might be selecting overloaded brokers.
example) threshold =-15, {b1=25, b2=100, b3=100, b4=100, b5=100} => b2, b3, b4, b5 will be selected.

And another edge case is that with a low variance of the usage distribution, it might be selecting none of the brokers.
example) threshold =10, {b1=24, b2=25, b3=26} => none will be selected.

If we want to care more about the bigger randomization pool, one could just randomly select any brokers below LoadBalancerBrokerOverloadedThresholdPercentage, which could be just another strategy.

Reply:
Truely. An algorithm that can be applied to the many cases is good, but cannot cover all cases.
> Another edge case is that with the negative threshold, it might be selecting overloaded brokers.
example) threshold =-15, {b1=25, b2=100, b3=100, b4=100, b5=100} => b2, b3, b4, b5 will be selected.

The operator need to know the status of the cluster and configure the threshold appropriately.
- negative threshold applies when a large number of brokers have less load. 
- positive threshold applies when a large number of brokers have heavy load.

> And another edge case is that with a low variance of the usage distribution, it might be selecting none of the brokers.
example) threshold =10, {b1=24, b2=25, b3=26} => none will be selected.

Assign randomly if none is selected.
<img width="854" alt="image" src="https://user-images.githubusercontent.com/4970972/177025193-55f818ca-d5a4-4e2c-b192-39590dbbfe5d.png">

Reply:
It is good we have a fallback case that just randomly selects any broker, when empty.

But threshold=-10 {b1:1, b2:25, b3:25,b4=25 ... b10=25}, then

Reply:
I prefer just keep the algorithm simple, otherwise it is hard to debug and understand.

The `diffThreshold` might filter all candidates out and finally just randomly select the broker from all candidates, which might be high load.

The principle of selection is to let the broker with lower load be selected, such as selecting a broker which is under avg load.

Reply:
It would be great if we could explain more about the negative and positive threshold considerations in the config `broker.conf`.

Otherwise, LGTM.